crates/metering/src/block.rs

```diff
@@ -0,0 +1,109 @@
+use alloy_primitives::B256;
+use eyre::{Result as EyreResult, eyre};
+use reth::revm::db::State;
+use reth_evm::{ConfigureEvm, execute::BlockBuilder};
+use reth_optimism_chainspec::OpChainSpec;
+use reth_optimism_evm::{OpEvmConfig, OpNextBlockEnvAttributes};
+use reth_optimism_primitives::OpBlock;
+use reth_primitives_traits::{Block as BlockT, SealedHeader};
+use reth_provider::{HashedPostStateProvider, StateRootProvider};
+
+use crate::types::{MeterBlockResponse, MeterBlockTransactions};
+
+/// Re-executes a block and meters execution time, state root calculation time, and total time.
+///
+/// Takes a state provider at the parent block, the chain spec, and the block to meter.
+///
+/// Returns `MeterBlockResponse` containing:
+/// - Block hash
+/// - EVM execution time for all transactions
+/// - State root calculation time
+/// - Total time
+/// - Per-transaction timing information
+pub fn meter_block<SP>(
+    state_provider: SP,
+    chain_spec: Arc<OpChainSpec>,
+    block: &OpBlock,
+    parent_header: &SealedHeader,
+) -> EyreResult<MeterBlockResponse>
+where
+    SP: reth_provider::StateProvider + StateRootProvider + HashedPostStateProvider,
+{
+    let block_hash = block.header().hash_slow();
+    let block_number = block.header().number();
+    let transactions: Vec<_> = block.body().transactions().cloned().collect();
+    let tx_count = transactions.len();
+
+    // Create state database from parent state
+    let state_db = reth::revm::database::StateProviderDatabase::new(&state_provider);
+    let mut db = State::builder().with_database(state_db).with_bundle_update().build();
+
+    // Set up block attributes from the actual block header
+    let attributes = OpNextBlockEnvAttributes {
+        timestamp: block.header().timestamp(),
+        suggested_fee_recipient: block.header().beneficiary(),
+        prev_randao: block.header().mix_hash().unwrap_or(B256::random()),
+        gas_limit: block.header().gas_limit(),
+        parent_beacon_block_root: block.header().parent_beacon_block_root(),
+        extra_data: block.header().extra_data().clone(),
+    };
+
+    // Execute transactions and measure time
+    let mut transaction_times = Vec::with_capacity(tx_count);
+
+    let evm_start = Instant::now();
+    {
+        let evm_config = OpEvmConfig::optimism(chain_spec);
+        let mut builder = evm_config.builder_for_next_block(&mut db, parent_header, attributes)?;
+
+        builder.apply_pre_execution_changes()?;
+
+        for tx in &transactions {
+            let tx_start = Instant::now();
+            let tx_hash = tx.tx_hash();
+
+            // Recover the signer to create a Recovered transaction for execution
+            let signer = tx
+                .recover_signer()
+                .map_err(|e| eyre!("Failed to recover signer for tx {}: {}", tx_hash, e))?;
+            let recovered_tx =
+                alloy_consensus::transaction::Recovered::new_unchecked(tx.clone(), signer);
```

Comment:
I would expect signer recovery to not be part of the metered tx execution time. This is something that can always be parallelized, so we should probably record this separately.

Reply:
468eb75